### ISSUE TITLE
Simulation memory leaks when adding ATMForce.

### ISSUE THREAD

**Issue by 9527567** [openmm/openmm#4457]
This code has a memory leak. When I comment out `system.addForce(atmforce)`, it no longer occurs. 
@egallicc 
~~~python3
from openmm import *
from openmm.app import *
from openmm.unit import *


def test_atmforce(pdbtopfile, systemfile):
 
    pdb = PDBFile(pdbtopfile)


    with open(systemfile) as input:
        system = XmlSerializer.deserialize(input.read())


    dt = 0.002*picoseconds
    temperature = 300*kelvin
    friction = 1.0/picosecond


    ubcore = 0.0 * kilocalorie_per_mole
    acore = 0.0
    lmbd = 0.0
    lambda1 = lmbd
    lambda2 = lmbd
    alpha = 0.0 / kilocalorie_per_mole
    uh = 0.0 * kilocalorie_per_mole
    w0coeff = 0.0 * kilocalorie_per_mole
    direction = 1.0
    umsc = 0.0 * kilocalorie_per_mole


    topology = pdb.topology

    atmforce = ATMForce(lambda1, lambda2,  alpha * kilojoules_per_mole, uh/kilojoules_per_mole, w0coeff/kilojoules_per_mole, umsc/kilojoules_per_mole, ubcore/kilojoules_per_mole, acore, direction )
    system.addForce(atmforce)
    integrator = LangevinMiddleIntegrator(temperature, friction, dt)
    simulation = Simulation(topology, system, integrator)

    
if __name__ == "__main__":
    while True:
        test_atmforce('10-4HBV.pdb', '10-4HBV_sys.xml')
~~~
[10-4HBV.zip](https://github.com/openmm/openmm/files/14441711/10-4HBV.zip)

**Comment by peastman** [comment]
What leads you to think it has a memory leak?  Can you describe exactly what you observe?

A memory leak means that memory is progressively lost, for example that total memory use steadily increases as a simulation runs.  But your script doesn't run a simulation.  It just creates some objects and then immediately exits.  What would it mean to leak memory in that case?

**Comment by 9527567** [comment]
Yes, this code caused a continuous increase in memory usage and was eventually killed due to running out of memory.

The above code was not simulated because I wanted to provide as little code as possible to locate the problem. Anyway, this code should not run out of memory.

**Comment by peastman** [comment]
Can you provide your complete code and all necessary input files?

**Comment by 9527567** [comment]
All input files are in [10-4HBV.zip](https://github.com/openmm/openmm/files/14455525/10-4HBV.zip). Unzip it and run test.py.
On my machine with 8GB of RAM, it will run out of memory and crash at the index of around 400.

**Comment by peastman** [comment]
You aren't running a simulation.  All your script does is keep creating objects and never doing anything with them.

8 GB is a very small amount of memory.  You're probably running right at the edge of your computer's memory, and some small fluctuation eventually pushes it over the edge.

**Comment by egallicc** [comment]
@9527567 are you investigating this because of some misbehavior of AToM-OpenMM? If so, we can continue the discussion there if it makes sense.

**Comment by 9527567** [comment]
> You aren't running a simulation. All your script does is keep creating objects and never doing anything with them.
> 
> 8 GB is a very small amount of memory. You're probably running right at the edge of your computer's memory, and some small fluctuation eventually pushes it over the edge.

Yes, 8GB is small, but the same thing happens on a 128GB machine, it just takes longer to occur on machines with larger memory capacity.
There is no need to run a time-consuming simulation that involves a large amount of code.

**Comment by 9527567** [comment]
> @9527567 are you investigating this because of some misbehavior of AToM-OpenMM? If so, we can continue the discussion there if it makes sense.

I don't know if this can be considered a problem with AToM OpenMM. Whenever a simulation is created and ATmforce is added, the memory grows very quickly, to the point where it occupies the machine's memory.

**Comment by peastman** [comment]
What platform are you using?  I can't tell from the script.  Can you try doing it with both the reference and CUDA platforms and see if it makes a difference?

**Comment by 9527567** [comment]
Reference and CUDA behavior are the same. If this script is not sufficient for problem observation, I will try to add as little code as possible to conduct the simulation, which may take some time.

**Comment by peastman** [comment]
I'll try running it through valgrind and see if it finds any memory problems.

**Comment by peastman** [comment]
You were right: there really was a memory leak.  The fix is in #4461.  Thanks for catching it!

**Comment by egallicc** [comment]
Thank you
